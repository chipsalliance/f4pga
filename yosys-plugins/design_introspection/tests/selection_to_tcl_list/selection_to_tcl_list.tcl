yosys -import
if { [info procs selection_to_tcl_list] == {} } { plugin -i design_introspection }
yosys -import  ;# ingest plugin commands

proc selection_to_tcl_list_through_file { selection } {
    set file_name [test_output_path "[pid].txt"]
    select $selection -write $file_name
    set fh [open $file_name r]
    set result [list]
    while {[gets $fh line] >= 0} {
	lappend result $line
    }
    close $fh
    file delete $file_name
    return $result
}

proc test_selection { rfh selection } {
    if {[expr {[selection_to_tcl_list_through_file $selection] != [selection_to_tcl_list $selection]}]} {
    	puts "List from file: [selection_to_tcl_list_through_file $selection]"
    	puts "List in selection: [selection_to_tcl_list $selection]"
	error "Test with selection: $selection failed"
    } else {
	puts $rfh [selection_to_tcl_list $selection]
    }
}

read_verilog $::env(DESIGN_TOP).v
read_verilog -specify -lib -D_EXPLICIT_CARRY +/xilinx/cells_sim.v
read_verilog -lib +/xilinx/cells_xtra.v
hierarchy -check -auto-top

# Test the selection command and write results to file
set rfh [open [test_output_path "selection_to_tcl_list.txt"] w]

set selection_tests [list "t:*" "w:*" "*"]
foreach test $selection_tests {
    test_selection $rfh $test
}

close $rfh
