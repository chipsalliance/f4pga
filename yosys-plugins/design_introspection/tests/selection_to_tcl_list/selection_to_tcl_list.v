module top (
    input clk,
    input clk2,
    input [1:0] in,
    output [5:0] out
);

  reg [1:0] cnt = 0;
  wire clk_int_1, clk_int_2;
  IBUF ibuf_proxy (
      .I(clk),
      .O(ibuf_proxy_out)
  );
  IBUF ibuf_inst (
      .I(ibuf_proxy_out),
      .O(ibuf_out)
  );
  assign clk_int_1 = ibuf_out;
  assign clk_int_2 = clk_int_1;

  always @(posedge clk_int_2) begin
    cnt <= cnt + 1;
  end

  middle middle_inst_1 (
      .clk(ibuf_out),
      .out(out[2])
  );
  middle middle_inst_2 (
      .clk(clk_int_1),
      .out(out[3])
  );
  middle middle_inst_3 (
      .clk(clk_int_2),
      .out(out[4])
  );
  middle middle_inst_4 (
      .clk(clk2),
      .out(out[5])
  );

  assign out[1:0] = {cnt[0], in[0]};
endmodule

module middle (
    input  clk,
    output out
);

  reg [1:0] cnt = 0;
  wire clk_int;
  assign clk_int = clk;
  always @(posedge clk_int) begin
    cnt <= cnt + 1;
  end

  assign out = cnt[0];
endmodule
