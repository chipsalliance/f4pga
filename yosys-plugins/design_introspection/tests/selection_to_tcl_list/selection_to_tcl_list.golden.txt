{middle/$add$selection_to_tcl_list.v:54$5} top/middle_inst_4 top/middle_inst_3 top/middle_inst_2 top/middle_inst_1 {top/$add$selection_to_tcl_list.v:22$2} top/ibuf_inst top/ibuf_proxy
{middle/$1\cnt[1:0]} {middle/$add$selection_to_tcl_list.v:54$5_Y} {middle/$0\cnt[1:0]} middle/clk_int middle/cnt middle/out middle/clk {top/$1\cnt[1:0]} {top/$add$selection_to_tcl_list.v:22$2_Y} {top/$0\cnt[1:0]} top/ibuf_out top/ibuf_proxy_out top/clk_int_2 top/clk_int_1 top/cnt top/out top/in top/clk2 top/clk
{middle/$1\cnt[1:0]} {middle/$add$selection_to_tcl_list.v:54$5_Y} {middle/$0\cnt[1:0]} middle/clk_int middle/cnt middle/out middle/clk {middle/$add$selection_to_tcl_list.v:54$5} {middle/$proc$selection_to_tcl_list.v:50$6} {middle/$proc$selection_to_tcl_list.v:53$4} {top/$1\cnt[1:0]} {top/$add$selection_to_tcl_list.v:22$2_Y} {top/$0\cnt[1:0]} top/ibuf_out top/ibuf_proxy_out top/clk_int_2 top/clk_int_1 top/cnt top/out top/in top/clk2 top/clk top/middle_inst_4 top/middle_inst_3 top/middle_inst_2 top/middle_inst_1 {top/$add$selection_to_tcl_list.v:22$2} top/ibuf_inst top/ibuf_proxy {top/$proc$selection_to_tcl_list.v:8$3} {top/$proc$selection_to_tcl_list.v:21$1}
